### PR TITLE
Add `BaseSeedable` class.

### PR DESCRIPTION
**Pull request by marcofavorito**
## Proposed changes

This is the base class for all objects that need to keep a random state.

In particular, the interface provides the following features:
- it is able to accept a random seed from the client;
- it allows the client reset the random seed;
- it provides a getter to the internal random generator;
- it allows to sample a random seed (in the range [0, 2^32 - 1]).

## Fixes

n/a

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [X] I have read the [CONTRIBUTING](../blob/master/CONTRIBUTING.md) doc
- [X] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works

## Further comments

n/a

### CODE REVIEW

File: black_it/utils/seedable.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    @random_state.setter
+    def random_state(self, random_state: Optional[int]) -> None:
+        """Set the random state."""
+        self._set_random_state(random_state)
+
+    def _set_random_state(self, random_state: Optional[int]) -> None:
+        """Set the random state, private use."""
+        self.__random_state = random_state
+        self.__random_generator = default_rng(self.random_state)
```
Comment (AldoGl) [inline black_it/utils/seedable.py:67]: maybe this is a naive comment, but shouldn't this be `self.__random_generator = default_rng(random_state)`? If we use `self.random_state` in the argument I am not sure we would get the behaviour we want

**Review by AldoGl — APPROVED** [review]
LGTM, I provide just a couple of comments

File: black_it/utils/seedable.py (hunk 1 above)
Comment (marcofavoritobi) [reply black_it/utils/seedable.py:67]: `self.random_state` is the property getter, that reads from `self.__random_state`, which is set the line above

File: black_it/utils/seedable.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+    def __init__(
+        self,
+        random_state: Optional[int] = None,
+    ) -> None:
+        """
+        Initialize the sampler.
+
+        Args:
+            random_state: the internal state of the sampler, fixing this numbers the sampler behaves deterministically
```
[the remark this answers is not in the document]
Comment (marcofavoritobi) [reply black_it/utils/seedable.py:46]: OK Thanks

[the remark this answers is not in the document]
Comment (marcofavoritobi) [reply black_it/utils/seedable.py:46]: (It could be a calibrator too)

[the remark this answers is not in the document]
Comment (AldoGl) [reply black_it/utils/seedable.py]: right!

File: black_it/utils/seedable.py (hunk 1 above)
Comment (AldoGl) [reply black_it/utils/seedable.py:68]: Ok, this makes sense, thanks

**Pull request merged**